[Bug] Enable noisy sims in pulser backend

Roland-djee opened the pull request:
Closes #421.

dominikandreasseitz commented on qadence/backends/pulser/backend.py:
```diff
@@ -235,13 +236,13 @@ def run_dm(
         noise_probs = noise.options.get("noise_probs", None)
-        if noise_probs is None:
-            KeyError(f"A range of noise probabilies should be passed. Got {noise_probs}.")
+        if noise_probs is None or not isinstance(noise_probs, Iterable):
```
i thought noise is a dataclass?

dominikandreasseitz replied:
ah, `noise_probs` sorry. are there any requirements for the shape of the noise?

Roland-djee replied:
What do you mean by "shape of the noise" ?

Roland-djee replied:
On second thoughts, `noise_probs` might either be a single float value or a container of floats. Will rectify this.

dominikandreasseitz reviewed (approved):
lgtm, just a few general questions regarding noise

Roland-djee commented:
@rajaiitp would you be able to review ?

rajaiitp commented on qadence/backends/pulser/backend.py:
```diff
@@ -229,19 +230,23 @@ def run_dm(
+            KeyError("A `noise probs` option should be passed to the <class QuantumModel>.")
+        if not (isinstance(noise_probs, float) or isinstance(noise_probs, Iterable)):
+            KeyError(
+                "A single or a range of noise probabilies"
```
```suggestion
                "A single or a range of noise probabilities"
```

rajaiitp commented on tests/conftest.py:
```diff
@@ -22,6 +23,47 @@
 
 
+@fixture
+def noisy_pulser_sim() -> Tensor:
```
The test cases are not uniform, we have density matrix comparison for one and the expectation value for the other. Why is it done this way?  
Also are we just storing the output from actual simulation and comparing with it again ? The test case was  just to check if the workflow goes through smoothly. Is the assertion check in the end important ?

Roland-djee replied:
Addressed in #427

rajaiitp commented on qadence/backends/pulser/backend.py:
```diff
@@ -250,7 +255,14 @@ def run_dm(
                 sequence = self.assign_parameters(circuit, param_values_el)
                 sim_result = simulate_sequence(sequence, self.config, state)
```
Get the final statevector here, flip the order like in line 218 and then append it. This would match the description similar to the run method. I think then you need to change line 52 in PSRExepctation to not redo computing the final statevector there. 
Feel free to ignore this if you think its irrelevant. Just trying to get things to be uniform. Not sure where the exchange of 0,1 convention is included in this density matrix flow

Roland-djee commented on qadence/backends/pulser/backend.py:
```diff
@@ -306,14 +307,26 @@ def expectation(
+                    noise=noise,
+                    param_values=param_values,
+                    state=state,
+                    endianness=endianness,
```
Addressed in #427

Roland-djee commented on qadence/backends/pulser/backend.py:
```diff
@@ -235,13 +236,13 @@ def run_dm(
 
         # Pulser requires numpy types.
-        for noise_prob in noise_probs.numpy():
+        for noise_prob in noise_probs:
```
Addressed in #427

Roland-djee commented:
@dominikandreasseitz @rajaiitp I messed up locally with this branch. I re-opened a new PR #427 that addresses all your points above. Will take it from there.
